Add documentation for OneDrive (#36768)
home-assistant/home-assistant.github.io: source/_integrations/onedrive.markdown

---
title: OneDrive
description: Instructions on how to setup OneDrive to be used with backups.
ha_release: 2025.2
ha_category:
  - Backup
ha_iot_class: Cloud Polling
ha_config_flow: true
ha_domain: onedrive
ha_codeowners:
  - '@zweckj'

> ha_integration_type: service

Reviewer: ```suggestion
ha_integration_type: service
related:
  - docs: /common-tasks/general/#backups
    title: Backups
```

---

> This integration allows you to use [OneDrive](https://www.microsoft.com/en-us/microsoft-365/onedrive/online-cloud-storage) for use with Home Assistant Backups.

Reviewer: ```suggestion
This integration allows you to use [OneDrive](https://www.microsoft.com/en-us/microsoft-365/onedrive/online-cloud-storage) for use with [Home Assistant Backups](/common-tasks/general/#backups).
```

{% include integrations/config_flow.md %}
{% configuration_basic %}
Client ID:
  description: "Application Id of the app registration to be used with the integration. Uses Home Assistant provided by default."
Client secret:
  description: "Application secret for the app registration. Uses Home Assistant provided by default."

{% endconfiguration_basic %}

{% include integrations/option_flow.md %}

## Requested permissions by the integration

The integration will request the following permissions on your OneDrive for the integration to work:

- `Files.ReadWrite.AppFolder`: Grants the application permission to read and write in its own, app-specific folder inside your OneDrive
- `offline_access`: Grants the application permission to refresh its authentication token without requiring your manual intervention
- `openid`: Grants the application permission to read basic information, e.g. if you have a OneDrive

## Getting application credentials

This integration comes with a predefined set of [application credentials](https://www.home-assistant.io/integrations/application_credentials/) through Home Assistant account linking. 

> Everything is done in a delegated user context, so nobody will ever have access to your data except you. However, if you want to use your own credentials, 

Reviewer: what do you mean by "is done in a delegated user context"?

Author: The app does not have permission to do anything on its own (application permissions) but only in user context (with a signed in user - delegated permissions)

> Everything is done in a delegated user context, so nobody will ever have access to your data except you. However, if you want to use your own credentials, 

Reviewer: ```suggestion
Nobody will ever have access to your data except you, as the app does not have permission to do anything on its own (application permission). It only works with a signed in user. However, if you want to use your own credentials, 
```
ah, I see. maybe something like this to give more context?

---
title: OneDrive
description: Instructions on how to setup OneDrive to be used with backups.
ha_release: 2025.2
ha_category:
  - Backup
ha_iot_class: Cloud Polling
ha_config_flow: true
ha_domain: onedrive
ha_codeowners:
  - '@zweckj'
ha_integration_type: service
related:
  - docs: /common-tasks/general/#backups
    title: Backups
---

> This integration allows you to use [OneDrive](https://www.microsoft.com/en-us/microsoft-365/onedrive/online-cloud-storage) for use with [Home Assistant Backups](/common-tasks/general/#backups).

Reviewer 2: ```suggestion
This integration allows you to use [OneDrive](https://www.microsoft.com/en-us/microsoft-365/onedrive/online-cloud-storage) for [Home Assistant Backups](/common-tasks/general/#backups).
```

Backups will be created in a folder called `home-assistant.io\backups_<id>` in the `App Folder` of your OneDrive.

> `id` is part of your Home Assistant instance's unique id to allow backups from multiple instances to the same OneDrive.

Reviewer 2: ```suggestion
`id` is part of your Home Assistant instance's unique id to allow backups from multiple instances to the same OneDrive account.
```

The integration only has access to an application specific `home-assistant.io` folder in the `App Folder` and cannot access to any other parts of your OneDrive.

{% note %}

> Because of an issue in the Graph API, the application-specific folder sometimes is called `Graph` instead of `home-assistant.io`.

Reviewer 2: is there any linked issue for this? also, does this mean that other apps could have access to HA backup if they endup getting the `Graph` folder too?

Author: issue yes - https://github.com/OneDrive/onedrive-api-docs/issues/1866. Cross access no, it's not done by name and will just generate multiple folders `Graph (1)` etc.

Author: added the link to the docs

{% endnote %}

{% include integrations/config_flow.md %}
{% configuration_basic %}
Client ID:

>   description: "Application ID of the app registration to be used with the integration. Uses Home Assistant provided by default."

Reviewer 2: I tried this and the field is mandatory and had no default value:
![image](https://github.com/user-attachments/assets/e1bed7ff-41bd-49ef-a358-8983f26fa2d6)

Author: yes, because it's currently not stored in HA cloud (this can only be done after it's merged according to frenck)

Author: @abmantis should work now

Client secret:
  description: "Application secret for the app registration. Uses Home Assistant provided by default."

{% endconfiguration_basic %}

> 
> {% include integrations/option_flow.md %}

Reviewer 2: ```suggestion
```
does the integration have options flow?

Author: nope

---
title: OneDrive
description: Instructions on how to setup OneDrive to be used with backups.
ha_release: 2025.2
ha_category:
  - Backup
ha_iot_class: Cloud Polling
ha_config_flow: true
ha_domain: onedrive
ha_codeowners:
  - '@zweckj'
ha_integration_type: service
related:
  - docs: /common-tasks/general/#backups
    title: Backups
---

This integration allows you to use [OneDrive](https://www.microsoft.com/en-us/microsoft-365/onedrive/online-cloud-storage) for use with [Home Assistant Backups](/common-tasks/general/#backups).

Backups will be created in a folder called `home-assistant\backups_<id>` in the `App Folder` of your OneDrive.
`id` is part of your Home Assistant instance's unique id to allow backups from multiple instances to the same OneDrive.
The integration only has write access to an application specific `home-assistant` folder in the `App Folder` and cannot write to any other parts of your OneDrive.

{% note %}

> Because of an issue in the Graph API the application specific folder sometimes is called `Graph` instead of `home-assistant`.

Reviewer: ```suggestion
Because of an issue in the Graph API, the application-specific folder sometimes is called `Graph` instead of `home-assistant`.
```

{% endnote %}

{% include integrations/config_flow.md %}
{% configuration_basic %}
Client ID:

>   description: "Application Id of the app registration to be used with the integration. Uses Home Assistant provided by default."

Reviewer: ```suggestion
  description: "Application ID of the app registration to be used with the integration. Uses Home Assistant provided by default."
```

Client secret:
  description: "Application secret for the app registration. Uses Home Assistant provided by default."

{% endconfiguration_basic %}

{% include integrations/option_flow.md %}

## Requested permissions by the integration

The integration will request the following permissions on your OneDrive for the integration to work:

- `Files.ReadWrite.AppFolder`: Grants the application permission to read and write in its own, app-specific folder inside your OneDrive
- `Files.Read`: Grant permission to read your OneDrive and its files. Required to get your OneDrive.
- `offline_access`: Grants the application permission to refresh its authentication token without requiring your manual intervention
- `openid`: Grants the application permission to read basic information, e.g. if you have a OneDrive

<img src='/images/integrations/onedrive/onedrive-permissions.png' alt='Lists of permissions that the application will request.'>

## Getting application credentials

This integration comes with a predefined set of [application credentials](https://www.home-assistant.io/integrations/application_credentials/) through Home Assistant account linking. 

> Nobody will ever have access to your data except you, as the app does not have permission to do anything on its own. It only works with a signed in user (it only has `delegated` not `application permissions`). 

Reviewer: ```suggestion
Nobody will ever have access to your data except you, as the app does not have permission to do anything on its own. It only works with a signed-in user (it only has `delegated` not `application permissions`). 
```

However, if you want to use your own credentials, follow [this guide](https://learn.microsoft.com/en-us/entra/identity-platform/quickstart-register-app?tabs=certificate) to create your own client ID and secret.

{% note %}
You will need an Azure tenant with an active Azure subscription to create your own client credentials.
{% endnote %}

> ## Known Limitations

Reviewer: ```suggestion
## Known limitations
```

> - Only personal OneDrives are supported at the moment

Reviewer: ```suggestion
- Only personal OneDrives are supported at the moment.
```
